lib/crystalball/map_generator.rb

```diff
@@ -28,48 +28,58 @@ def build(config)
       def default_config
         {
           execution_detector: ExecutionDetector.new(Dir.pwd),
-          map_class: StandardMap,
-          map_storage: MapStorage::YAMLStorage.new(Pathname('execution_map.yml'))
+          map_storage: MapStorage::YAMLStorage.new(Pathname('execution_map.yml')),
+          dump_threshold: 100
         }
       end
     end
 
-    def initialize(execution_detector:, map_class:, map_storage:)
+    def initialize(execution_detector:, map_storage:, dump_threshold:)
       @execution_detector = execution_detector
       @map_storage = map_storage
-      @map_class = map_class
+      @dump_threshold = dump_threshold.to_i
     end
 
     def start!
       raise 'Repository is not pristine! Please stash all your changes' unless repo.pristine?
 
       self.map = nil
       map_storage.clear!
+      map_storage.dump(map.metadata.to_h)
     end
 
     def refresh_for_case(example)
       before = Coverage.peek_result
       example.run
       after = Coverage.peek_result
 
-      map.stash(CaseMap.new(example, execution_detector.detect(before, after)))
+      map << CaseMap.new(example, execution_detector.detect(before, after))
+
+      check_dump_threshold
     end
 
     def finalize!
-      map.dump
+      map_storage.dump(map.cases) if map.size.positive?
     end
 
     def map
-      @map ||= map_class.new(map_storage, metadata: {commit: repo.object('HEAD').sha})
+      @map ||= ExecutionMap.new(metadata: {commit: repo.object('HEAD').sha})
     end
 
     private
 
-    attr_reader :execution_detector, :map_storage, :map_class
+    attr_reader :execution_detector, :map_storage, :dump_threshold
     attr_writer :map
 
     def repo
       @repo ||= GitRepo.new('.')
     end
+
+    def check_dump_threshold
+      return unless dump_threshold.positive? && map.size >= dump_threshold
```

Comment:
I think `dump_storage: nil` would be better when user doesn't want dump storage, rather than setting it to 0 or -1

Reply:
`nil` will also work since we do `.to_i`